pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1998,43 +1998,68 @@ protected void handleEndTxnOnPartition(CommandEndTxnOnPartition command) {
-                            "Topic " + topic + " is not found."));
-                    return;
-                }
-                optionalTopic.get().endTxn(txnID, txnAction, command.getTxnidLeastBitsOfLowWatermark())
+        CompletableFuture<Optional<Topic>> topicFuture = service.getTopicIfExists(TopicName.get(topic).toString());
+        topicFuture.thenAccept(optionalTopic -> {
+            if (optionalTopic.isPresent()) {
+                optionalTopic.get().endTxn(txnID, txnAction, lowWaterMark)
                         .whenComplete((ignored, throwable) -> {
                             if (throwable != null) {
-                                log.error("Handle endTxnOnPartition {} failed.", topic, throwable);
+                                log.error("handleEndTxnOnPartition fail!, topic {}, txnId: [{}], "
+                                        + "txnAction: [{}]", topic, txnID, TxnAction.valueOf(txnAction), throwable);
                                 ctx.writeAndFlush(Commands.newEndTxnOnPartitionResponse(
                                         requestId, BrokerServiceException.getClientErrorCode(throwable),
-                                        throwable.getMessage()));
+                                        throwable.getMessage(),
+                                        txnID.getLeastSigBits(), txnID.getMostSigBits()));
                                 return;
                             }
                             ctx.writeAndFlush(Commands.newEndTxnOnPartitionResponse(requestId,
                                     txnID.getLeastSigBits(), txnID.getMostSigBits()));
                         });
-            });
-        } else {
-            log.error("handleEndTxnOnPartition faile ! The topic {} does not exist in broker, "
-                    + "txnId: [{}], txnAction: [{}]", topic, txnID, TxnAction.valueOf(txnAction));
-            ctx.writeAndFlush(Commands.newEndTxnOnSubscriptionResponse(
-                    requestId, txnID.getLeastSigBits(), txnID.getMostSigBits(),
-                    ServerError.ServiceNotReady,
-                    "The topic " + topic + " is not exist in broker."));
-        }
+
+            } else {
+                getBrokerService().getManagedLedgerFactory()
+                        .asyncExists(TopicName.get(topic).getPersistenceNamingEncoding())
+                        .thenAccept((b) -> {
+                            if (b) {
+                                log.error("handleEndTxnOnPartition fail ! The topic {} does not exist in broker, "
+                                                + "txnId: [{}], txnAction: [{}]", topic,
+                                        txnID, TxnAction.valueOf(txnAction));
+                                ctx.writeAndFlush(Commands.newEndTxnOnPartitionResponse(requestId,
+                                        ServerError.ServiceNotReady,
+                                        "The topic " + topic + " is not exist in broker.",
```

Comment:
there is a typo, what about:  "The topic " + topic + " does not exist in broker."